ocaml/xenopsd/c_stubs/xenctrlext_stubs.c

```diff
@@ -394,6 +409,127 @@ CAMLprim value stub_xenctrlext_cputopoinfo(value xch)
 	CAMLreturn(result);
 }
 
+CAMLprim value stub_xenforeignmemory_open(value logger, value open_flags)
+{
+        CAMLparam2(logger, open_flags);
+        struct xentoollog_logger *log_handle = NULL;
+        struct xenforeignmemory_handle *fmem;
+        CAMLlocal1(result);
+
+        if(Is_some(logger)) {
+                log_handle = Xtl_val(Some_val(logger));
+        }
+
+        fmem = xenforeignmemory_open(log_handle, Int_val(open_flags));
+
+        if(fmem == NULL) {
+                caml_failwith("Error when opening foreign memory handle");
+        }
+
+        result = caml_alloc(1, Abstract_tag);
+        Xfm_val(result) = fmem;
+
+        CAMLreturn(result);
+}
+
+CAMLprim value stub_xenforeignmemory_close(value fmem)
+{
+        CAMLparam1(fmem);
+        int retval;
+
+        retval = xenforeignmemory_close(Xfm_val(fmem));
+
+        if(retval < 0) {
+                caml_failwith("Error when closing foreign memory handle");
+        }
```

Comment:
To protect again double close you could also then set Xfm_val(fmem) to NULL here after the call succeeds. Looking at the implementation of `xenforeignmemory_close` it'll just exist when its arg is null.
We could also explicitly test against NULL before calling it and raise Invalid_argument if it is NULL.